RecoLocalTracker/SiPixelRecHits/src/PixelCPEGenericBase.cc

```diff
@@ -0,0 +1,237 @@
+                                         const SiPixelLorentzAngle* lorentzAngleWidth = nullptr)
+    : PixelCPEBase(conf, mag, geom, ttopo, lorentzAngle, genErrorDBObject, nullptr, lorentzAngleWidth, 0),
+      edgeClusterErrorX_{static_cast<float>(conf.getParameter<double>("EdgeClusterErrorX"))},
+      edgeClusterErrorY_{static_cast<float>(conf.getParameter<double>("EdgeClusterErrorY"))},
+      useErrorsFromTemplates_{conf.getParameter<bool>("UseErrorsFromTemplates")},
+      truncatePixelCharge_{conf.getParameter<bool>("TruncatePixelCharge")},
+      xerr_barrel_l1_{convertDoubleVecToFloatVec(conf.getParameter<std::vector<double>>("xerr_barrel_l1"))},
+      yerr_barrel_l1_{convertDoubleVecToFloatVec(conf.getParameter<std::vector<double>>("yerr_barrel_l1"))},
+      xerr_barrel_ln_{convertDoubleVecToFloatVec(conf.getParameter<std::vector<double>>("xerr_barrel_ln"))},
+      yerr_barrel_ln_{convertDoubleVecToFloatVec(conf.getParameter<std::vector<double>>("yerr_barrel_ln"))},
+      xerr_endcap_{convertDoubleVecToFloatVec(conf.getParameter<std::vector<double>>("xerr_endcap"))},
+      yerr_endcap_{convertDoubleVecToFloatVec(conf.getParameter<std::vector<double>>("yerr_endcap"))},
+      xerr_barrel_l1_def_{static_cast<float>(conf.getParameter<double>("xerr_barrel_l1_def"))},
+      yerr_barrel_l1_def_{static_cast<float>(conf.getParameter<double>("yerr_barrel_l1_def"))},
+      xerr_barrel_ln_def_{static_cast<float>(conf.getParameter<double>("xerr_barrel_ln_def"))},
+      yerr_barrel_ln_def_{static_cast<float>(conf.getParameter<double>("yerr_barrel_ln_def"))},
+      xerr_endcap_def_{static_cast<float>(conf.getParameter<double>("xerr_endcap_def"))},
+      yerr_endcap_def_{static_cast<float>(conf.getParameter<double>("yerr_endcap_def"))} {};
+
+std::unique_ptr<PixelCPEBase::ClusterParam> PixelCPEGenericBase::createClusterParam(const SiPixelCluster& cl) const {
+  return std::make_unique<ClusterParamGeneric>(cl);
+}
+
+//-----------------------------------------------------------------------------
+//!  Collect the edge charges in x and y, in a single pass over the pixel vector.
+//!  Calculate charge in the first and last pixel projected in x and y
+//!  and the inner cluster charge, projected in x and y.
+//-----------------------------------------------------------------------------
+void PixelCPEGenericBase::collect_edge_charges(ClusterParam& theClusterParamBase,  //!< input, the cluster
+                                               int& q_f_X,                         //!< output, Q first  in X
+                                               int& q_l_X,                         //!< output, Q last   in X
+                                               int& q_f_Y,                         //!< output, Q first  in Y
+                                               int& q_l_Y,                         //!< output, Q last   in Y
+                                               bool truncate) {
```

Comment:
this method was const in `PixelCPEGeneric`.
```suggestion
                                               bool truncate) const {
```
I didn't notice member data modifications. Did I miss something?

Reply:
This is a `static` member function. It could be `non-static` `const` too, whichever is preferable.

Reply:
ah, nevermind then.
it's OK as is.